[Feature 7] Generate Graphical Summary

katarinaavucic opened the pull request:
Finished the classes needed to generate the graphical summary. The GUI for the graphs looks a little janky, I am thinking of adding a new issue to improve the interface later!

All functions are documented with Javadocs.

Included tests for GenerateSummaryUCI and GenerateSummaryUCInterpretor, to ensure proper functionality of the classes. Currently, everything is included in the tests besides the GenerateSummaryC. To see what the graphs look like, you can uncomment the while loop in the GenerateSummaryUCITest.

Additionally, I have modified the build.gradle file to import both packages needed to create the graphs (jcommon and jfreechart). The GenerateSummaryUCITest is failing right now due to an issue with gradle because these packages are not imported, so the file is not able to build. It is working correctly on my computer, but to be safe we might want to wait until the tutorial to merge this branch.

Please let me know if any code needs further Javadoc clarification or more testing!

katarinaavucic committed:
Added the GenerateSummaryC, GenerateSummaryID, and GenerateSummaryIB classes with Javadocs

katarinaavucic committed:
Added the GenerateSummaryUCI and GenerateSummaryUCInterpretor classes with Javadocs, and modified GenerateSummaryC and GenerateSummaryIB classes

katarinaavucic committed:
Added the GenerateSummaryOB, GenerateSummaryOD and GenerateSummaryP classes with Javadocs, and modified the build.gradle file with new packages for plotting

katarinaavucic committed:
Added the GenerateSummaryUCInterpreterTest to ensure that the class works for full and empty months

katarinaavucic committed:
Modified GenerateSummaryUCI to allow for a negative remainder (user spending over their budget)

katarinaavucic committed:
Modified GenerateSummaryP to allow for a negative remainder (user spending over their budget, made a mistake on naming of last commit)

katarinaavucic committed:
Modified GenerateSummaryUCITest to test for all possible cases (a full month, an empty month, and an error in MonthMenuV)

katarinaavucic committed:
Fixed the Javadocs.

katarinaavucic committed:
Pushed build.gradle again.

katarinaavucic committed:
Updated to a newer version of JFreeChart.

YinLingithub commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryOB.java:
```diff
@@ -0,0 +1,15 @@
+     * outputData to contain the resulting JPanel.
+     * @param outputData a GenerateSummaryOD object holding the formatted data needed to generate the graphs
+     */
+    void formatOutputData(GenerateSummaryOD outputData);
```
This should return GenerateSummaryOD.

katarinaavucic replied:
Okay!

YinLingithub commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryOD.java:
```diff
@@ -0,0 +1,78 @@
+    public Map<String, ArrayList<Double>> getStatisticalData() {
+        return statisticalData;
+    }
+
```
Your output data should not contain Swing components (i.e. JPanel), so any method below this comment should not be in this class.

katarinaavucic replied:
Thanks, I am refactoring the code now!

YinLingithub commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCI.java:
```diff
@@ -0,0 +1,39 @@
+        GenerateSummaryOD outputData = new GenerateSummaryOD(interpreter.getRemainder(), interpreter.getStatisticalData());
+        presenter.formatOutputData(outputData);
+        return outputData;
+    }
```
Should directly return presenter.formatOutputData(outputData).

YinLingithub commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCInterpreter.java:
```diff
@@ -0,0 +1,67 @@
+        return this.remainder;
+    }
+
+}
```
I feel like the constructor and getter methods are doing the same thing. Consider moving these code to a private helper method in GenerateSummaryUCInteractor (so you don't need getters to make a duplicate copy of your statistical data).

katarinaavucic replied:
Okay, that's a good point! Initially, I was concerned about having the constructor method too long I like entirely forgot private methods exist haha.

YinLingithub replied:
I saw your new implementation for the GenerateSummaryUCInteractor, and it does make more sense now, so it's fine as a helper class!

YinLingithub commented on src/main/java/views/generate_summary_views/GenerateSummaryP.java:
```diff
@@ -0,0 +1,104 @@
+        this.statisticalData = outputData.getStatisticalData();
+        outputData.setBarGraphPanel(generateBarGraph());
+        outputData.setPieChartPanel(generatePieChart());
+    }
```
This method should return GenerateSummaryOD object, which will be passed all the way to your view. Your view can have the methods below as private helper methods, and it will need a frame to hold all outputs from these methods.

katarinaavucic replied:
Okay, that will be in my next commit when I refactor the code! Thanks for figuring out the issue :).

YinLingithub commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCI.java:
```diff
@@ -0,0 +1,39 @@
+        GenerateSummaryOD outputData = new GenerateSummaryOD(interpreter.getRemainder(), interpreter.getStatisticalData());
+        presenter.formatOutputData(outputData);
+        return outputData;
+    }
```
You are throwing EntityException, but never catching them? You should probably catch them in your view.

katarinaavucic replied:
Yes, I had planned to catch EntityException in the MonthMenuV, since that is the View that the GenerateSummaryC will be called in. The code will try the controller, and is EntityException is raised it will call the displayErrorMessage method in the GenerateSummaryV

YinLingithub replied:
Yeah, what would be possible, but it would also violate Single Responsibility Principle, so we have to have a separate view.

aricasas replied:
This is resolved now right? Since you are catching it and displaying an error message in GenerateSummaryV

katarinaavucic committed:
Refactored code so that graphs are generated in the GenerateSummaryV. Also, made all generation of statisticalData and the remainder in the constructor of the GenerateSummaryUCI and created getters for retrieving that information.

katarinaavucic committed:
Delete GenerateSummaryUCITest.java

Removed GenerateSummaryUCITest to (hopefully) eliminate the auto-grading issue.

katarinaavucic committed:
Added the @throws EntityException in the Javadoc of the GenerateSummaryC.

katarinaavucic committed:
Deleted GenerateSummaryUCITest.java

katarinaavucic committed:
Merge remote-tracking branch 'origin/4-feature-7-generate-graphical-summary' into 4-feature-7-generate-graphical-summary

katarinaavucic committed:
Modified some Javadocs and spacing and optimized imports. Also got rid of old methods in the GenerateSummaryOD.

katarinaavucic committed:
Modified more Javadocs and updated the UCI so that it creates the outputData (not the presenter).

katarinaavucic committed:
Modified the GenerateSummaryV so that everything is executed in the view. Also added aditional tests to GenerateSummaryUCInterpreterTest, if the month is full.

katarinaavucic committed:
Minor typo in GenerateSummaryV Javadoc.

aricasas commented on src/main/java/views/generate_summary_views/GenerateSummaryV.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public GenerateSummaryV(GenerateSummaryC controller, SessionStorage session, int monthID){
+        JFrame screen = new JFrame("Graphical Summary");
+        screen.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
```
This call is unneeded since you call screen.setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE); at the bottom

katarinaavucic replied:
Oops I didn't even notice that, thanks for catching it!

aricasas commented on src/main/java/views/generate_summary_views/GenerateSummaryV.java:
```diff
@@ -0,0 +1,133 @@
+            screen.add(graphs);
+            graphs.add(generateBarGraph(statisticalData));
+            graphs.add(generatePieChart(remainder, statisticalData));
+        } catch (Exception e) {
```
Could you change this to the specific exception type expected

katarinaavucic replied:
Yup! Initially I had just a general exception just to catch everything, but on second thoughts it isn't neccessary to do that.

aricasas commented on src/main/java/views/generate_summary_views/GenerateSummaryV.java:
Good helper methods to build JPanels!

katarinaavucic replied:
Thank you!

aricasas commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryID.java:
```diff
@@ -0,0 +1,39 @@
+public class GenerateSummaryID {
+
+    SessionStorage currentSession;
+    int monthID;
```
Should these be private final?

katarinaavucic replied:
Yeah, thanks for catching that!

aricasas commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryIB.java:
```diff
@@ -0,0 +1,17 @@
+     * Generates a graphical representation of the MonthlyStorage data associated with monthID in the currentSession.
+     * @param inputData a GenerateSummaryID object holding the SessionStorage and monthID needed to retrieve data
+     * @return a JPanel holding the graphs
+     */
```
This JavaDoc is outdated

katarinaavucic replied:
I just fixed all these errors in the latest commit!

aricasas commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCInterpreter.java:
This new implementation of the Interpreter is a lot easier to read and understand at a glance. Nice job!

katarinaavucic replied:
Thank you!! :D

aricasas reviewed (changes requested):
Overall great! I just had some minor comments but everything seems to be done now

katarinaavucic committed:
Fixed duplicate call and specific exception in GenerateSummaryV. Made attributes private final in GenerateSummaryID. Fixed outdated Javadoc in GenerateSummaryIB.

aricasas reviewed (dismissed):
Everything looks good!

YinLingithub reviewed (dismissed):
Very concise error message in view! Everything looks clear.

luluwuwu commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCInterpreter.java:
```diff
@@ -0,0 +1,103 @@
+     * @return a Map holding data used to make the graphs
+     */
+    private Map<String, ArrayList<Double>> generateStatisticalData(MonthlyStorage monthlyStorage, Map<Category,
+            ArrayList<Expense>> mappedExpenses){
```
Does this mean that you're also generating a graph for "Other" Category? I thought it does not have a budget. What will you be displaying then?

luluwuwu commented on src/main/java/use_cases/generate_summary_use_case/GenerateSummaryUCInterpreter.java:
```diff
@@ -0,0 +1,103 @@
+     * @return a Map holding data used to make the graphs
+     */
+    private Map<String, ArrayList<Double>> generateStatisticalData(MonthlyStorage monthlyStorage, Map<Category,
+            ArrayList<Expense>> mappedExpenses){
```
Does this mean that you're also generating a graph for "Other" Category? I thought it does not have a budget. What will you be displaying then?

katarinaavucic replied:
The "Other" category has a default budget of 0! So on the graph, the money they spend in the "Other" category is shown, but there is not a remainder shown.

katarinaavucic replied:
Uh oh I just realized my bar graphs are wrong :O. Thanks for mentioning this I literally would not have noticed!!

luluwuwu reviewed (commented):
Just wondering how a specific case of Category is displayed.

katarinaavucic committed:
Fixed the GenerateSummaryUCInterpreter so that statisticalData holds the remainder of money left in the budget and not the budget. This solves the issue of adding the spent and budget together in the graphs (and not the spent and remainder).

aricasas reviewed (approved):
Everything looks good

luluwuwu reviewed (approved):
Looking good

YinLingithub reviewed (approved):
Looks great!
